Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+   * are not changed.
+   * \param[in] mpi_env - The MPI environment used for communication.
+   * \param[in] local_version - The local NdFlattener structure.
+   */
+  template<typename MPI_Environment_type>
+  void refresh( 
+    MPI_Environment_type const& mpi_env,
+    Base const* local_version 
+  ) {
+    Index* Nodes_all_0 = nullptr;
+    int* displs = new int[mpi_env.size]; // {0, 1, ..., size-1}
+    int* ones = new int[mpi_env.size]; // {1,1, ...}
+    for(int r=0; r<mpi_env.size; r++){
+      displs[r] = r;
+      ones[r] = 1;
+    }
+    LowestLayer::count_g(mpi_env, &Nodes_all_0, static_cast<LowestLayer const*>(local_version), displs, ones);
+    LowestLayer::set_g(mpi_env, &Nodes_all_0, static_cast<LowestLayer const*>(local_version));
+
+    delete[] Nodes_all_0; // allocated by count_g
+    delete[] displs;
+    delete[] ones;
+  }
+
+protected:
+  /*! \brief Count the distributed flatteners' numbers of nodes, and set nNodes.
+   *
+   * \param[in] mpi_env - MPI environment for communication
+   * \param[out] Nodes_all - [k][r] is set to number of nodes in layer (k+1), rank r.
+   * \param[in] local_version - local instance to be send to the other processes
+   * \param[in] displs - {0,1,...,size-1}
+   * \param[in] ones - {1,1,...,1}
+   */
+  template<typename MPI_Environment_type>
+  void count_g(MPI_Environment_type const& mpi_env,
+         Index** Nodes_all,
+         CurrentLayer const* local_version,
+         int const* displs, int const* ones )
+  { 
+    assert( Nodes_all[K-1]==nullptr);
+    Nodes_all[K-1] = new Index[mpi_env.size];
+    nNodes = 0;
+    // gather numbers of nodes in the current layer from all processes
+    mpi_env.MPI_Allgatherv( &(local_version->nNodes), 1, mpi_env.mpi_index, Nodes_all[K-1], ones, displs, mpi_env.mpi_index, mpi_env.comm );
+    for(int r=0; r<mpi_env.size; r++){
+      nNodes += Nodes_all[K-1][r];
+    }
+    Base::count_g(mpi_env, Nodes_all, static_cast<const Base*>(local_version), displs, ones);
+  }
+
+  /*! \brief Gather the distributed flatteners' data and index arrays into the allocated arrays.
+   *
+   * \param[in] mpi_env - MPI environment for communication
+   * \param[in] Nodes_all - [k][r] is the number of nodes in layer (k+1), rank r.
+   * \param[in] local_version - local instance to be send to the other processes
+   */
+  template<typename MPI_Environment_type>
+  void set_g(MPI_Environment_type const& mpi_env,
+         Index** Nodes_all,
+         CurrentLayer const* local_version )
+  { 
+
+    int* Nodes_all_K_as_int = new int[mpi_env.size];
+    int* Nodes_all_k_cumulated = new int[mpi_env.size+1]; // [r] is number of nodes in the current layer, summed over all processes with rank below r 
+    // plus one. Used as displacements in Allgatherv, but we do not want to transfer the initial zeros and rather the last element of indices, 
+    // which is the local nNodes of the layer below. Note that MPI needs indices of type 'int'.
```

Comment:
basically the counts and displacements of Allgatherv

Reply:
yes